### ISSUE TITLE
Resolve related object from dbcontext in Field conflicted with scope lifetime dependency injection

### ISSUE THREAD

**Issue by diouxis** [graphql-dotnet/graphql-dotnet#1184]
## Description

Here is an example of article graph type I created in the system
``` csharp
public ArticleGraphType(UserRepository users)
{
    Name = "Article";

    Id(p => p.ArticleId);
    Field(p => p.Title);
    Field<UserGraphType>()
        .Name("User")
        .ResolveAsync(async ctx => await users.Get(ctx.Source.UserId));
}
```
But I get exception when I query it, and it says the DbContext has been disposed, I did a quick debug, I can see it is true. All the scope dependencies are get disposed as soon as finish the resolve in the QueryGraphType, so it can't use continue to use them to resolve object in the other ObjectGraphType.

Could someone help me check if it is an issue in the library or some mistake I made?

**Comment by sungam3r** [comment]
See #1151 and #1160

**Comment by furier** [comment]
If the GraphType has the same lifetime as the Service, it should be fine.

We used to have all GraphTypes as Scoped per request and the same with all Services, then constructor injection was fine. However that means the entire graph is created on every request so we moved to GraphTypes as Singletons, only injecting Singletons in the constructor and Scoped services are resolved per resolver method, with IServiceProvider RequestServices.

**Comment by sungam3r** [comment]
Using RequestServices for resolving dependencies is ServiceLocator antipattern.

**Comment by furier** [comment]
> Using RequestServices for resolving dependencies is ServiceLocator antipattern.

I agree but it's the only way we have found that makes it possible to keep the `GraphQLSchema` as a singleton, and it vastly improved performance.

**Comment by sungam3r** [comment]
It is not the only way. See #1160

**Comment by furier** [comment]
the implementation uses the RequestServices provider from the HttpContext, so what is the big difference? why is the other way an antipattern over this one?

**Comment by sungam3r** [comment]
Because this is **implementation** and not user code. No need to burden the user code with the responsibility of manually resolving dependencies. Also `HttpContext` is just one of the possible existing providers that is hidden from user code. User code (resolvers) should not be tied to ASP.NET infrastructure at all.

**Comment by furier** [comment]
Firstly, I am not here to argue with you @sungam3r, this is an educational discussion for me.

> User code (resolvers) should not be tied to ASP.NET infrastructure at all.

Fully agree on this one, but can be solved by having my own `IDependencyResolver` interface where I have an implementation resolving dependencies with an `IServiceProvider` which in this case happens to be RequestServices. Thus my user code is not tied to ASP.NET's infrastructure.

However the `GraphTypes` I define are very much user code? I defined those types, I define what constructor arguments (services) are needed to resolve fields, the only difference to me is that manually resolving the dependencies in the resolvers may lead to more lines of code if the dependencies are used in multiple fields.

**Comment by sungam3r** [comment]
Ok.

>  the only difference to me is that manually resolving the dependencies in the resolvers may lead to more lines of code if the dependencies are used in multiple fields.

Not only more lines but **hidden** dependencies. Think about how configure/test/mock it (not only by you as author who know/remember all the code).

**Comment by sungam3r** [comment]
Closed as duplicate in favor of #648 and related issues.
